### ISSUE TITLE
Security concerns

### ISSUE THREAD

**Issue by Falci** [namebasehq/handshake-id-manager#2]
Hey!

Namebase contributes a lot with the Handshake community, mostly bring tools that reduces the friction for adoption.

I can't say I'm happy with Namebase's participation in OSS side, but this repo seems a good step in the right direction.

I was reading the source code and security model strategy used here has raised some flags.
I'm not sure if those are temporary or final decisions.

1. `Math.random()` is not cryptographically secure.
[Mozilla](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random):
> Math.random() does not provide cryptographically secure random numbers. Do not use them for anything related to security.

https://github.com/namebasehq/handshake-id-manager/blob/84846be634c7e764df9e5436d087eaf49b2dcbec/src/providers/IdentitiesContextProvider/IdentitiesContextProvider.svelte#L37

More concerning than that is following line:

2. Don't use `localStorage` for sensitive data
https://github.com/namebasehq/handshake-id-manager/blob/84846be634c7e764df9e5436d087eaf49b2dcbec/src/providers/IdentitiesContextProvider/IdentitiesContextProvider.svelte#L38
https://github.com/namebasehq/handshake-id-manager/blob/84846be634c7e764df9e5436d087eaf49b2dcbec/src/services/IdentityService/index.ts#L29-L42

In a nutshell: this script generates a secure enough pair of private/public keys, put it a vault but let the vault's key in its door.

The user experience is really smooth, but at which cost?

**Comment by brandondees** [comment]
I hadn't looked at the details yet, but yes these two details are unacceptable for secure authentication use. Thanks for taking a closer look @Falci

**Comment by turbomaze** [comment]
Thanks so much for reviewing, quickly:

1. We didn't want to ever generate passphrases for users, we wanted to require the user to specify their own. You asked about temporary/final decisions and this one was certainly temporary. We want to get feedback as early as we can before committing to more extensive user flows.
2. With a sufficiently secure passphrase you could even store your encrypted keys somewhere public like Sia.

I don't think too much has been shared about how we intended this to be used (we plan to release more docs on this soon), so let me explain:

This is a basic example of an "identity manager" that you can use for Handshake login. You'll be able to specify which identity manager you'd like to use by creating a text record like:

```
_idmanager.johnxu.      1167    IN      TXT     v=0;url=http://localhost:3000
```

In the login flow, the authentication server will redirect you to your specified identity manager, which is responsible for signing the challenge and redirecting you back. I _strongly_ encourage anyone that cares about their security to fork this repo and host their own on a URL that they control. You can then modify the key storage and passphrase logic as much as you'd like!

For my personal use, I would want an identity manager that gives me a button to copy paste the gpg command I need to run to sign the challenge on the isolated, non-networked VM that contains my PGP keys. Finally, I'd copy paste the armored signature back into the identity manager, which would format it into the correct shape for Handshake login.

You can even imagine building an identity manager that incorporates something like webauthn and your fingerprint, etc. We thought a lot about what flow would be the least limiting for users so that they could have full control over their security. What you see in this repo is a first example of a hyper-convenient flow that gave us something tangible to test out these ideas.

@Falci @brandondees @jeefave

**Comment by rithvikvibhu** [comment]
Ooh nice I wanna make it work with yubikeys (U2F, I don't have the latest one with webauthn :/). Instead of forking this complete repo, it would be nice to see it as a separate library that can be loaded and used (for sinpapeles or this, whatever).

I realize it's still in early stages, but it will help keep all implementations up to date instead of relying on everyone to manually update heavily modified forks. I think the main interfaces are fairly ready?

**Comment by Falci** [comment]
@turbomaze Thanks for the quick reply.

That sounds amazing. I'm definitely going to provide my own id manager. Looking forward to the docs.

**Comment by turbomaze** [comment]
> Ooh nice I wanna make it work with yubikeys (U2F, I don't have the latest one with webauthn :/). Instead of forking this complete repo, it would be nice to see it as a separate library that can be loaded and used (for sinpapeles or this, whatever).
> 
> I realize it's still in early stages, but it will help keep all implementations up to date instead of relying on everyone to manually update heavily modified forks. I think the main interfaces are fairly ready?

Yeah this is a good point, if you have thoughts on how a library system could work for this please share. I expect some security conscious people will have hesitations about this particular repo since it uses svelte which means there's a somewhat opaque compilation process happening. The main interface is extremely simple, just need to parse out the challenge and redirect URL. 

As I'm writing, one interesting option I'm imagining is supporting a bunch of different auth options in this main identity manager, and switching on which gets used depending on the specific name being authenticated. When a name is first being configured, the user can select the key management strategy they'd like to use. This can be saved in the DNS record.

**Comment by rithvikvibhu** [comment]
I'm no expert on building libraries, esp. those focused on security, so probably someone with more xp should start one (or I'll try to in a few weeks). Just some things could think of:
- Should only accept challenge, manage strategies (below), return a response - no redirection or action taken by itself
- Should be able to generate key pairs, fingerprint, TXT record
- Should not contain any UI related code
- May handle DNS querying to check for records, match with existing profiles, etc.
<br>

> supporting a bunch of different auth options in this main identity manager

The multi-strategy idea sounds good. I like how Passport.js works. Anyone can write a strategy (for Google, FB, OAuth, Openid, etc.) by inheriting an [abstract class](https://github.com/jaredhanson/passport-strategy) and all authentication servers (ID managers in this case) can use it.

I don't see why a name has to be anchored to a single strategy.
- The same key is stored in 2 or more places (logical OR between a yubikey and a browser extension so won't need yubikey on personal computers, but will work smoothly in a public setting)
- Can require multiple strategies to sign the challenge (logical AND, like multi-factor with a yubikey and a passphrase).

P.S. since ID managers are completely browser-side, they can be hosted anywhere (Vercel, GitHub, Netlify, S3, etc.) In the future, a "Deploy to Netlify" (etc.) button will make it easy for everyone to "spin up" their own ID Managers.

**Comment by turbomaze** [comment]
> * The same key is stored in 2 or more places (logical OR between a yubikey and a browser extension so won't need yubikey on personal computers, but will work smoothly in a public setting)
> * Can require multiple strategies to sign the challenge (logical AND, like multi-factor with a yubikey and a passphrase).

Agree with everything you wrote above except this bit -- for cryptography you want to avoid moving/duplicating key material. So each device would need its own key. Currently Handshake login generates a new device ID for every device you set your name up on, but I think it's currently configured so that one name <-> one identity manager. That 1 identity manager could certainly support multiple strategies though! After being sent to your identity manager you can select whether you'd like to use your offline PGP key to sign, or your fingerprint, etc.

**Comment by damajor** [comment]
> 2. With a sufficiently secure passphrase you could even store your encrypted keys somewhere public like Sia.

I don't agree with this one, even with long and/or complex encryption and passwords, the secret should never be stored online and publicly available.

---

Regarding the identity manager:
> `_idmanager.johnxu.      1167    IN      TXT     v=0;url=http://localhost:3000`

Why not using the IN SRV records ? I think they better match the use case of puclicly exposing how to contact various services, they are just made for that purpose.

**Comment by turbomaze** [comment]
> I don't agree with this one, even with long and/or complex encryption and passwords, the secret should never be stored online and publicly available.

Thanks for participating in this discussion. Some thoughts:

Security is measured in $$ to compromise, which we can estimate by comparing three common factors: 1) something you have, 2) something you are, and 3) something you know. Stealing something you have (e.g. hardware key) requires kidnapping/robbery (expensive), and impersonating something you are (e.g. faking an iris/fingerprint scan) is also pretty difficult (though not impossible). Which brings us to 3, passcodes/words/phrases etc.

Let's assume the auth challenge signature cannot be man-in-the-middled and submitted in a race against your legitimate auth request (all 3 of these factors are equally susceptible to this so it's not an interesting consideration).

With this assumption, then in the extreme case where your passphrase is as high entropy as a private key, guessing the password is ~equivalent to guessing the private key in the hardware device. It doesn't matter if the encrypted private key your passphrase is securing is stored on a flash drive, in local storage, on Sia, or on a billboard. If you specify your desired security in $$, you can compute the cost to brute force an n bit passphrase and select one exactly long enough to satisfy your needs (it'll end up being a lot shorter than a private key).

If you're really important or paranoid then maybe your $$ security requirement is high enough to include secret NSA retroreflectors being installed in your keyboard, opening up an additional exfiltration path for your passphrase when you type it in. But if your $$ requirement is this high, then storing your encrypted key somewhere else besides Sia/local storage isn't going to make a difference. There's not much you can do besides moving the signing step to an isolated, secure platform (e.g. a hardware key).

Summary: sufficiently high entropy passphrases offer a considerable amount of security. Users exist on a spectrum of $$ security requirements and the vast majority are perfectly served by a high-entropy passphrase. Those with higher requirements should not at all rely on a passphrase and instead opt for more secure factors.

> Why not using the IN SRV records ? I think they better match the use case of puclicly exposing how to contact various services, they are just made for that purpose.

This is a good idea, thanks. We chose TXT records for flexibility. In this issue we've talked about supporting multiple auth strategies in the identity manager. This might be best stored as additional metadata in the text record, might be hard to represent in SRV as the scope grows. I can imagine storing the hash of the identity manager source code in the TXT record and installing a browser extension that ensures the identity manager you're redirected to matches, etc.
